File: 1.Stack_Queue/4949.py
import sys
sinput = sys.stdin.readline

while True:
    st = sinput()
    aa = ''
    stack = list()
    if st==".":
        break

    for i in st:
        if i in "()[]":
            aa += i

    for i in aa:
        if i in "([":
            stack.append(i)
        else:
            if i == ")":
                if len(stack) > 0 and stack[-1] == "(":
                    stack.pop()
                else:
                    stack.append(i)
                    break
            else:
                if len(stack) > 0 and stack[-1] == "[":
                    stack.pop()
                else:
                    stack.append(i)
                    break
        
    if stack:
        print('no')
    else:
        print('yes')

File: barkingDog series/0x08_stack2/4949.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    while (true) {
        string inp;
        getline(cin, inp);

        if (inp == ".") {
            break;
        }

        stack<char> stk;
        bool flag = true;
        
        for (auto t : inp) {
            if (t == '[' || t == '(') {
                stk.push(t);
            } else if (t == ']') {
                if (!stk.empty() && stk.top() == '[') {
                    stk.pop();
                } else {
                    flag = false;
                    break;
                }
            } else if (t == ')') {
                if (!stk.empty() && stk.top() == '(') {
                    stk.pop();
                } else {
                    flag = false;
                    break;
                }
            } else {
                continue;
            }
        }
        
        if (flag || stk.empty()) {
            cout << "yes" << "\n";
        } else {
            cout << "no" << "\n";
        }
    }
    
    return 0;
}